File: libs/auth-utils/permissions.ts
import { Permission, PERMISSION_MATRIX, Role, ROLE_HIERARCHY } from './constants';

export const ROLES = {
  VIEWER: 'viewer',
  ENGINEER: 'engineer',
  TEAM_LEAD: 'team_lead',
  ADMIN: 'admin',
} as const;

export function hasRoleAtLeast(role: Role, minimum: Role): boolean {
  return ROLE_HIERARCHY.indexOf(role) >= ROLE_HIERARCHY.indexOf(minimum);
}

export function canAccess(permission: Permission, roles: readonly string[]): boolean {
  const allowed = PERMISSION_MATRIX[permission];
  return roles.some((role) => allowed.includes(role as Role));
}

File: libs/auth-utils/permissions.py
from .constants import PERMISSION_MATRIX, ROLE_HIERARCHY


def has_role_at_least(role: str, minimum: str) -> bool:
    return ROLE_HIERARCHY.index(role) >= ROLE_HIERARCHY.index(minimum)


def can_access(permission: str, roles: list[str]) -> bool:
    allowed = PERMISSION_MATRIX.get(permission, [])
    return any(role in allowed for role in roles)
